mkosi/config.py

```diff
@@ -4904,13 +4907,22 @@ def finalize(self) -> dict[str, Any]:
         return ns
 
 
-def have_history(args: Args) -> bool:
-    return (
-        args.directory is not None
-        and args.verb.needs_build()
-        and ((args.verb != Verb.build and not args.force) or args.rerun_build_scripts)
-        and Path(".mkosi-private/history/latest.json").exists()
-    )
+def have_history(args: Args, tools: bool = False) -> bool:
+    filename = "tools" if tools else "latest"
+
+    if args.directory is None:
+        return False
+
+    if args.verb in (Verb.build, Verb.cat_config):
+        return False
+
+    if args.verb.needs_tools() and args.force > 0:
+        return False
+
+    if args.verb.needs_build() and args.force > 0 and not args.rerun_build_scripts:
+        return False
```

Comment:
The lower one cannot be hit, because if `args.verb.needs_build()` return true and `args.force` is non-zero, the upper one will already have returned `False`.

Reply:
Is this right? The upper one uses needs_tools() and the lower one uses needs_build().